discover/controller.go

```diff
@@ -0,0 +1,220 @@
+package discover
+
+import (
+	"reflect"
+	"time"
+
+	wraperrors "github.com/pkg/errors"
+	"k8s.io/api/apps/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kenjones-cisco/dapperdox/discover/models"
+)
+
+// watcher defines an event loop.
+//
+// The watcher guarantees the following consistency requirement: registry
+// view in the watcher is as AT LEAST as fresh as the moment notification
+// arrives, but MAY BE more fresh (e.g. "delete" cancels an "add" event).  For
+// example, an event for a service creation will see a service registry without
+// the service if the event is immediately followed by the service deletion
+// event.
+//
+// Handlers execute on the single worker queue in the order they are appended.
+// Handlers receive the notification event and the associated object.  Note
+// that all handlers must be appended before starting the watcher.
+type watcher interface {
+	// AppendServiceHandler notifies about changes to the service catalog.
+	AppendServiceHandler(f func(*models.Service, models.Event)) error
+
+	// AppendDeploymentHandler notifies about change to the deployment catalog.
+	AppendDeploymentHandler(f func(*models.Deployment, models.Event)) error
```

Comment:
since we never return an error, it would make sense to remove the return parameter

Reply:
removed